File: ruby/NumericDamage.rb
#encoding:utf-8

require_relative 'Damage'
require_relative 'NumericDamageToUI'

module Deepspace
    class NumericDamage < Damage

        public_class_method :new

        def initialize(nw,s)
            super(s)
            @nWeapons = nw
        end

        attr_reader :nWeapons

        def hasNoEffect
            return (@nWeapons == 0 && super)
        end

        def copy
            return NumericDamage.new(@nWeapons,nShields)
        end
      
        def getUIversion
            return NumericDamageToUI.new(self)
        end

        def adjust(w,s)
            #shmin = [@nShields,s.length].min ##aqui se pondria super?¿?¿?
            shmin = super(s)
            weapmin = [@nWeapons,w.length].min
      
            return NumericDamage.new(weapmin,shmin)
        end

        def discardWeapon(w)
            if @nWeapons > 0
                @nWeapons -= 1
            end
        end

        def to_s
            return "Numeric Damage. NShields: #{@nShields}, NWeapons: #{@nWeapons}"
        end
    end
end

File: java/Deepspace/src/deepspace/NumericDamage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author alvaro
 */
public class NumericDamage extends Damage{
    private int nWeapons;
    
    NumericDamage(int w, int s){
        super(s);
        nWeapons = w;
    }
    
    NumericDamage(NumericDamage danio){
        super(danio.getNShields());
        nWeapons = danio.nWeapons;
    }
    
    public int getNWeapons(){
        return nWeapons;
    }
    
    @Override
    public NumericDamage copy(){
        return new NumericDamage(nWeapons, getNShields());
    }
    
    @Override
    public void discardWeapon(Weapon w){
        if(nWeapons > 0)
            nWeapons--;
    }
    
    @Override
    public boolean hasNoEffect(){
        return ((getNShields() == 0) && (nWeapons == 0));
    }
    
    @Override
    public Damage adjust (ArrayList<Weapon>w, ArrayList<ShieldBooster>s){
        int shmin, weapmin;
        weapmin = Math.min(this.nWeapons, w.size());
        shmin = Math.min(getNShields(), s.size());
               
        return new NumericDamage(weapmin,shmin);  
    }
    
    @Override
    public NumericDamageToUI getUIversion(){
        return new NumericDamageToUI(this);
    }
    
    @Override
    public String toString(){
        String test;
        //CONTENT
        String nW = "- NWEAPONS: " + nWeapons + "\n";
     
        test = super.toString() + nW;
        
        return test;
    }
        
}
